### ISSUE TITLE
Assert on "this->IsDenseTensor()" failed: All tensors in the input TensorList must have the same shape and be densely packed.

### ISSUE THREAD

**Issue by HaoLIU94** [NVIDIA/DALI#3718]
Hello, 

I am trying to use DALI to do a test on the external source pipeline for object detection task.
While using _fn.external_source_, I would like to send labels which are bounding boxes. 
With that the size of labels could vary between images. I got the issue as indicated in the title. 

When I tried to put _fn.external_source(source=external_data, num_outputs=2,dtype=[types.UINT8, types.FLOAT], **batch=False**)_
It will not bother me with the size but instead show that _**'list' object has no attribute 'shape'**_

Here is the code that I used from your ExternalSource operator example

```python
batch_size = 3
epochs = 3

class ExternalInputIterator(object):
    def __init__(self, batch_size, device_id, num_gpus):
        self.images_dir = "../../data/images/"
        self.batch_size = batch_size
        with open(self.images_dir + "file_list.txt", 'r') as f:
            self.files = [line.rstrip() for line in f if line is not '']
        # whole data set size
        self.data_set_len = len(self.files)
        # based on the device_id and total number of GPUs - world size
        # get proper shard
        self.files = self.files[self.data_set_len * device_id // num_gpus:
                                self.data_set_len * (device_id + 1) // num_gpus]
        self.n = len(self.files)
    def __iter__(self):
        self.i = 0
        shuffle(self.files)
        return self

    def __next__(self):
        batch = []
        labels = []

        if self.i >= self.n:
            self.__iter__()
            raise StopIteration

        for _ in range(self.batch_size):
            jpeg_filename  = self.files[self.i % self.n]
            #######Here I try to put some artificial bounding boxes (x,y,h,w,cls)#######
            ####### and try to simulate each image could hava a different number of bbx
            if self.i%2:    
                label = [[0.5,0.3,0.2,0.2,8],[0.5,0.3,0.2,0.2,8]]
            else:
                label = [[0.5,0.3,0.2,0.2,8],[0.5,0.3,0.2,0.2,8],[0.5,0.3,0.2,0.2,8]]
            batch.append(np.fromfile(jpeg_filename, dtype = np.uint8))  # we can use numpy
            labels.append(torch.tensor([label])) # or PyTorch's native tensors
            self.i += 1
        return (batch, labels)

    def __len__(self):
        return self.data_set_len

    next = __next__

def ExternalSourcePipeline(batch_size, num_threads, device_id, external_data):
    pipe = Pipeline(batch_size, num_threads, device_id)
    with pipe:
        jpegs, labels = fn.external_source(source=external_data, num_outputs=2,dtype=[types.UINT8, types.FLOAT],batch=False)
        images = fn.decoders.image(jpegs, device="mixed")
        images = fn.resize(images, resize_x=640, resize_y=640)
        output = fn.cast(images, dtype=types.UINT8)
        pipe.set_outputs(output, labels)
    return pipe

eii = ExternalInputIterator(batch_size, 0, 1)

from nvidia.dali.plugin.pytorch import DALIGenericIterator as PyTorchIterator
from nvidia.dali.plugin.pytorch import LastBatchPolicy

pipe = ExternalSourcePipeline(batch_size=batch_size, num_threads=2, device_id = 0,
                              external_data = eii)
pii = PyTorchIterator(pipe,['image','labels'],last_batch_padded=True, last_batch_policy=LastBatchPolicy.PARTIAL)

for e in range(epochs):
    for i, data in enumerate(pii):
        print("epoch: {}, iter {}, real batch size: {}".format(e, i, len(data[0]["image"])))
    pii.reset()
```

**Comment by HaoLIU94** [comment]
Just noticed this 
https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.fn.box_encoder

**Comment by JanuszL** [comment]
Hi @HaoLIU94,

You need to either set `batch=True` as you return a whole batch at the time from your `ExternalInputIterator` or return only one sample at the time:
```
if self.i%2:
    label = [[0.5,0.3,0.2,0.2,8],[0.5,0.3,0.2,0.2,8]]
else:
    label = [[0.5,0.3,0.2,0.2,8],[0.5,0.3,0.2,0.2,8],[0.5,0.3,0.2,0.2,8]]
label = torch.tensor([label])
jpeg_filename  = self.files[self.i % self.n]
batch = np.fromfile(jpeg_filename, dtype = np.uint8)  # we can use numpy
self.i += 1
return (batch, label)```

**Comment by HaoLIU94** [comment]
Thanks you @JanuszL  for your quick response. 

It is a pity that if I can't use it with a batchsize bigger than 1. I found out that DALI does not support dynamique size for labels (here in my case bbx). 

If I am right, I have to construct my labels into anchors in the class **_ExternalInputIterator_** before send it to the 
 function **_ExternalSourcePipeline_** like what you do in _**nvidia.dali.fn.box_encoder**_ to make it a fixed tensor size otherwise I must send labels manually to GPUS without calling **_ExternalInputIterator_**. These two solutions would be sufficient I think.

**Comment by JanuszL** [comment]
Hi,

I'm not sure if I understand why you can't use it with a batch size bigger than 1. If you are talking about the requirement that DL FW iterator returns batch that is uniform then that is true. Otherwise, you can do it.

**Comment by HaoLIU94** [comment]
Hi Janusz,

I hope I can make it more clearer this time.

In class ExternalInputIterator
```python
class ExternalInputIterator(object):
```
If given batchsize > 1, the iterator will just return the number of items that we demand it.
For instance, when bs=2 you will get ([img1,img2],[label1,label2]). 
Because img1 and img2 they are different images, they may have different number of annotations (bbx)
from the test code above we will have: 
label1 = [[0.5,0.3,0.2,0.2,8],[0.5,0.3,0.2,0.2,8]]   with a size (2,5)
label2 = [[0.5,0.3,0.2,0.2,8],[0.5,0.3,0.2,0.2,8],[0.5,0.3,0.2,0.2,8]] with a size (3,5)

Apperently 
```python
fn.external_source(source=external_data, num_outputs=2,dtype=[types.UINT8, types.FLOAT],batch=True)
```
Does not like that. It will give me the error in the title. Bcz it wants the size of label1 = label2
especially when I put batchsize=1, there will be no such problems. 
It will not complain either when I have bs >1, with all the labels of uniform size.
Is it a concern of performance?

That's why I'm wondering how to do it when batchsize>1?
How to really take advantage of DALI to to batch training instead of single image each time.
like example here https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/general/data_loading/coco_reader.html
How it is done for nvidia.dali.fn.readers.coco() ?

**Comment by JanuszL** [comment]
You can use pad operator to make the batch uniform and return original shapes as an additional output to know what data can be discarded:
```
from nvidia.dali.pipeline import Pipeline
import nvidia.dali.fn as fn
import nvidia.dali.types as types
import torch
from random import shuffle
import numpy as np

batch_size = 3
epochs = 3

class ExternalInputIterator(object):
    def __init__(self, batch_size, device_id, num_gpus):
        self.images_dir = "docs/examples/data/images/"
        self.batch_size = batch_size
        with open(self.images_dir + "file_list.txt", 'r') as f:
            self.files = ["docs/examples/data/images/" + line.split()[0].rstrip() for line in f if line != '']
        # whole data set size
        self.data_set_len = len(self.files)
        # based on the device_id and total number of GPUs - world size
        # get proper shard
        self.files = self.files[self.data_set_len * device_id // num_gpus:
                                self.data_set_len * (device_id + 1) // num_gpus]
        self.n = len(self.files)
    def __iter__(self):
        self.i = 0
        shuffle(self.files)
        return self

    def __next__(self):
        batch = []
        labels = []

        if self.i >= self.n:
            self.__iter__()
            raise StopIteration

        for _ in range(self.batch_size):
            jpeg_filename  = self.files[self.i % self.n]
            #######Here I try to put some artificial bounding boxes (x,y,h,w,cls)#######
            ####### and try to simulate each image could hava a different number of bbx
            if self.i%2:
                label = [[0.5,0.3,0.2,0.2,8],[0.5,0.3,0.2,0.2,8]]
            else:
                label = [[0.5,0.3,0.2,0.2,8],[0.5,0.3,0.2,0.2,8],[0.5,0.3,0.2,0.2,8]]
            batch.append(np.fromfile(jpeg_filename, dtype = np.uint8))  # we can use numpy
            labels.append(torch.tensor([label])) # or PyTorch's native tensors
            self.i += 1
        return (batch, labels)

    def __len__(self):
        return self.data_set_len

    next = __next__

def ExternalSourcePipeline(batch_size, num_threads, device_id, external_data):
    pipe = Pipeline(batch_size, num_threads, device_id)
    with pipe:
        jpegs, labels = fn.external_source(source=external_data, num_outputs=2,dtype=[types.UINT8, types.FLOAT],batch=True)
        images = fn.decoders.image(jpegs, device="mixed")
        images = fn.resize(images, resize_x=640, resize_y=640)
        output = fn.cast(images, dtype=types.UINT8)
        shapes = fn.shapes(labels)
        labels = fn.pad(labels)
        pipe.set_outputs(output, labels, shapes)
    return pipe

eii = ExternalInputIterator(batch_size, 0, 1)

from nvidia.dali.plugin.pytorch import DALIGenericIterator as PyTorchIterator
from nvidia.dali.plugin.pytorch import LastBatchPolicy

pipe = ExternalSourcePipeline(batch_size=batch_size, num_threads=2, device_id = 0,
                              external_data = eii)
pii = PyTorchIterator(pipe,['image','labels', 'shapes'],last_batch_padded=True, last_batch_policy=LastBatchPolicy.PARTIAL)

for e in range(epochs):
    for i, data in enumerate(pii):
        print("epoch: {}, iter {}, real batch size: {}".format(e, i, len(data[0]["image"])))
    pii.reset()```
